From f7dbbdfe9f8ffdfe3d6823f4993733d74efd5ad6 Mon Sep 17 00:00:00 2001
From: Yorgos Thessalonikefs <yorgos@nlnetlabs.nl>
Date: Tue, 19 May 2026 14:09:28 +0100
Subject: [PATCH] Fix #1346: [FR] Please allow back TLS 1.2.

'tls-use-system-policy-versions' is introduced to allow Unbound to use
any system available TLS version when serving TLS.

---------

Co-authored-by: Wouter Wijngaards <wcawijngaards@users.noreply.github.com>
---
 daemon/daemon.c                |  3 ++-
 daemon/remote.c                |  3 ++-
 dnstap/unbound-dnstap-socket.c |  2 +-
 doc/example.conf.in            |  8 ++++++++
 doc/unbound.conf.5.in          | 29 ++++++++++++++++++++++++++++-
 doc/unbound.conf.rst           | 16 ++++++++++++++++
 util/config_file.c             |  3 +++
 util/config_file.h             |  2 ++
 util/configlexer.lex           |  1 +
 util/configparser.y            | 13 ++++++++++++-
 util/net_help.c                | 12 ++++++++----
 util/net_help.h                | 10 ++++++++--
 winrc/win_svc.c                |  6 ++++--
 13 files changed, 95 insertions(+), 13 deletions(-)

diff --git a/daemon/daemon.c b/daemon/daemon.c
index ea4e83e..6bfeede 100644
--- a/daemon/daemon.c
+++ b/daemon/daemon.c
@@ -215,7 +215,8 @@ setup_listen_sslctx(void** ctx, int is_dot, int is_doh,
 		cfg->tls_ciphers, cfg->tls_ciphersuites,
 		(cfg->tls_session_ticket_keys.first &&
 		cfg->tls_session_ticket_keys.first->str[0] != 0),
-		is_dot, is_doh, cfg->tls_protocols))) {
+		is_dot, is_doh, cfg->tls_protocols,
+		cfg->tls_use_system_policy_versions))) {
 		fatal_exit("could not set up listen SSL_CTX");
 	}
 }
diff --git a/daemon/remote.c b/daemon/remote.c
index 1eaf900..1998b53 100644
--- a/daemon/remote.c
+++ b/daemon/remote.c
@@ -153,7 +153,8 @@ remote_setup_ctx(struct daemon_remote* rc, struct config_file* cfg)
 		log_crypto_err("could not SSL_CTX_new");
 		return 0;
 	}
-	if(!listen_sslctx_setup(rc->ctx, cfg->tls_protocols)) {
+	if(!listen_sslctx_setup(rc->ctx, cfg->tls_protocols,
+		cfg->tls_use_system_policy_versions)) {
 		return 0;
 	}
 
diff --git a/dnstap/unbound-dnstap-socket.c b/dnstap/unbound-dnstap-socket.c
index 90b0f60..6fa6ca8 100644
--- a/dnstap/unbound-dnstap-socket.c
+++ b/dnstap/unbound-dnstap-socket.c
@@ -347,7 +347,7 @@ static struct tap_socket* tap_socket_new_tlsaccept(char* ip,
 	s->ev_cb = ev_cb;
 	s->data = data;
 	s->sslctx = listen_sslctx_create(server_key, server_cert, verifypem,
-		NULL, NULL, 0, 0, 0, tls_protocols);
+		NULL, NULL, 0, 0, 0, tls_protocols, 0);
 	if(!s->sslctx) {
 		log_err("could not create ssl context");
 		free(s->ip);
diff --git a/doc/example.conf.in b/doc/example.conf.in
index 63a8d65..19c3fe6 100644
--- a/doc/example.conf.in
+++ b/doc/example.conf.in
@@ -971,6 +971,14 @@ server:
 	# Changing the value requires a restart.
 	# tls-use-sni: yes
 
+	# Allow general-purpose version-flexible TLS server configuration that
+	# may be further restricted by the system's policy.
+	# Use only if you want to support legacy TLS client connections.
+	# Default is no and Unbound will only use the latest available TLS
+	# version.
+	# Changing the value requires a reload.
+	# tls-use-system-policy-versions: no
+
 	# TLS protocols.
 	# Changing the value requires a restart.
 	# tls-protocols: "TLSv1.2 TLSv1.3"
diff --git a/doc/unbound.conf.5.in b/doc/unbound.conf.5.in
index c5e19d9..368e3db 100644
--- a/doc/unbound.conf.5.in
+++ b/doc/unbound.conf.5.in
@@ -1306,7 +1306,34 @@ Default: yes
 .UNINDENT
 .INDENT 0.0
 .TP
-.B tls\-protocols: \fI\(dq<list of protocols>\(dq\fP 
+.B tls\-use\-system\-policy\-versions: \fI<yes or no>\fP
+Enable or disable general\-purpose version\-flexible TLS server configuration
+when serving TLS.
+This will allow the whole list of available TLS versions provided by the
+crypto library, which may have been further restricted by the system\(aqs
+crypto policy.
+.sp
+By default Unbound only uses the latest available TLS version.
+.sp
+\fBCAUTION:\fP
+.INDENT 7.0
+.INDENT 3.5
+Use only if you want to support legacy TLS client connections.
+.UNINDENT
+.UNINDENT
+.sp
+\fBNOTE:\fP
+.INDENT 7.0
+.INDENT 3.5
+Changing the value requires a reload.
+.UNINDENT
+.UNINDENT
+.sp
+Default: no
+.UNINDENT
+.INDENT 0.0
+.TP
+.B tls\-protocols: \fI\(dq<list of protocols>\(dq\fP
 Specify the allowed TLS protocol versions to use, in no particular order.
 Possible values are \fBTLSv1.2\fP and \fBTLSv1.3\fP\&.
 Enclose list of protocols in quotes (\fB\(dq\(dq\fP) and put spaces between them.
diff --git a/doc/unbound.conf.rst b/doc/unbound.conf.rst
index cc01451..e8bf308 100644
--- a/doc/unbound.conf.rst
+++ b/doc/unbound.conf.rst
@@ -1185,6 +1185,22 @@ These options are part of the ``server:`` section.
     Default: yes
 
 
+@@UAHL@unbound.conf@tls-use-system-policy-versions@@: *<yes or no>*
+    Enable or disable general-purpose version-flexible TLS server configuration
+    when serving TLS.
+    This will allow the whole list of available TLS versions provided by the
+    crypto library, which may have been further restricted by the system's
+    crypto policy.
+
+    By default Unbound only uses the latest available TLS version.
+
+    .. caution:: Use only if you want to support legacy TLS client connections.
+
+    .. note:: Changing the value requires a reload.
+
+    Default: no
+
+
 @@UAHL@unbound.conf@tls-protocols@@: *"<list of protocols>"*
     Specify the allowed TLS protocol versions to use, in no particular order.
     Possible values are ``TLSv1.2`` and ``TLSv1.3``.
diff --git a/util/config_file.c b/util/config_file.c
index adee403..6cb228b 100644
--- a/util/config_file.c
+++ b/util/config_file.c
@@ -132,6 +132,7 @@ config_create(void)
 	cfg->tls_cert_bundle = NULL;
 	cfg->tls_win_cert = 0;
 	cfg->tls_use_sni = 1;
+	cfg->tls_use_system_policy_versions = 0;
 	if(!(cfg->tls_protocols = strdup("TLSv1.2 TLSv1.3"))) goto error_exit;
 	cfg->https_port = UNBOUND_DNS_OVER_HTTPS_PORT;
 	if(!(cfg->http_endpoint = strdup("/dns-query"))) goto error_exit;
@@ -635,6 +636,7 @@ int config_set_option(struct config_file* cfg, const char* opt,
 	else S_STR("tls-ciphers:", tls_ciphers)
 	else S_STR("tls-ciphersuites:", tls_ciphersuites)
 	else S_YNO("tls-use-sni:", tls_use_sni)
+	else S_YNO("tls-use-system-policy-versions:", tls_use_system_policy_versions)
 	else if(strcmp(opt, "tls-protocols:") == 0) {
 		if(!cfg_tls_protocols_is_valid(val)) return 0;
 		free(cfg->tls_protocols);
@@ -1194,6 +1196,7 @@ config_get_option(struct config_file* cfg, const char* opt,
 	else O_STR(opt, "tls-ciphers", tls_ciphers)
 	else O_STR(opt, "tls-ciphersuites", tls_ciphersuites)
 	else O_YNO(opt, "tls-use-sni", tls_use_sni)
+	else O_YNO(opt, "tls-use-system-policy-versions", tls_use_system_policy_versions)
 	else O_STR(opt, "tls-protocols", tls_protocols)
 	else O_DEC(opt, "https-port", https_port)
 	else O_STR(opt, "http-endpoint", http_endpoint)
diff --git a/util/config_file.h b/util/config_file.h
index 6dadf0b..0a77478 100644
--- a/util/config_file.h
+++ b/util/config_file.h
@@ -148,6 +148,8 @@ struct config_file {
 	char* tls_ciphersuites;
 	/** if SNI is to be used */
 	int tls_use_sni;
+	/** if all TLS versions can be used; based on system policy (if any) */
+	int tls_use_system_policy_versions;
 	/** TLS protocols */
 	char* tls_protocols;
 
diff --git a/util/configlexer.lex b/util/configlexer.lex
index d6222cd..74caf69 100644
--- a/util/configlexer.lex
+++ b/util/configlexer.lex
@@ -261,6 +261,7 @@ tls-session-ticket-keys{COLON}	{ YDVAR(1, VAR_TLS_SESSION_TICKET_KEYS) }
 tls-ciphers{COLON}		{ YDVAR(1, VAR_TLS_CIPHERS) }
 tls-ciphersuites{COLON}		{ YDVAR(1, VAR_TLS_CIPHERSUITES) }
 tls-use-sni{COLON}		{ YDVAR(1, VAR_TLS_USE_SNI) }
+tls-use-system-policy-versions{COLON}		{ YDVAR(1, VAR_TLS_USE_SYSTEM_POLICY_VERSIONS) }
 tls-protocols{COLON}		{ YDVAR(1, VAR_TLS_PROTOCOLS) }
 https-port{COLON}		{ YDVAR(1, VAR_HTTPS_PORT) }
 http-endpoint{COLON}		{ YDVAR(1, VAR_HTTP_ENDPOINT) }
diff --git a/util/configparser.y b/util/configparser.y
index 6ea214e..cbea846 100644
--- a/util/configparser.y
+++ b/util/configparser.y
@@ -199,6 +199,7 @@ extern struct config_parser_state* cfg_parser;
 %token VAR_DISCARD_TIMEOUT VAR_WAIT_LIMIT VAR_WAIT_LIMIT_COOKIE
 %token VAR_WAIT_LIMIT_NETBLOCK VAR_WAIT_LIMIT_COOKIE_NETBLOCK
 %token VAR_STREAM_WAIT_SIZE VAR_TLS_CIPHERS VAR_TLS_CIPHERSUITES VAR_TLS_USE_SNI
+%token VAR_TLS_USE_SYSTEM_POLICY_VERSIONS
 %token VAR_TLS_PROTOCOLS
 %token VAR_IPSET VAR_IPSET_NAME_V4 VAR_IPSET_NAME_V6
 %token VAR_TLS_SESSION_TICKET_KEYS VAR_RPZ VAR_TAGS VAR_RPZ_ACTION_OVERRIDE
@@ -348,7 +349,8 @@ content_server: server_num_threads | server_verbosity | server_port |
 	server_stream_wait_size | server_tls_ciphers |
 	server_tls_ciphersuites | server_tls_session_ticket_keys |
 	server_answer_cookie | server_cookie_secret | server_ip_ratelimit_cookie |
-	server_tls_use_sni | server_edns_client_string | server_tls_protocols |
+	server_tls_use_sni | server_tls_use_system_policy_versions |
+	server_edns_client_string | server_tls_protocols |
 	server_edns_client_string_opcode | server_nsid |
 	server_zonemd_permissive_mode | server_max_reuse_tcp_queries |
 	server_tcp_reuse_timeout | server_tcp_auth_query_timeout |
@@ -1158,6 +1160,15 @@ server_tls_use_sni: VAR_TLS_USE_SNI STRING_ARG
 		free($2);
 	}
 	;
+server_tls_use_system_policy_versions: VAR_TLS_USE_SYSTEM_POLICY_VERSIONS STRING_ARG
+	{
+		OUTYY(("P(server_tls_use_system_policy_versions:%s)\n", $2));
+		if(strcmp($2, "yes") != 0 && strcmp($2, "no") != 0)
+			yyerror("expected yes or no.");
+		else cfg_parser->cfg->tls_use_system_policy_versions = (strcmp($2, "yes")==0);
+		free($2);
+	}
+	;
 server_tls_protocols: VAR_TLS_PROTOCOLS STRING_ARG
 	{
 		OUTYY(("P(server_tls_protocols:%s)\n", $2));
diff --git a/util/net_help.c b/util/net_help.c
index a0e23f5..d91204b 100644
--- a/util/net_help.c
+++ b/util/net_help.c
@@ -1226,13 +1226,15 @@ setup_ticket_keys_cb(void* sslctx)
 #endif /* HAVE_SSL */
 
 int
-listen_sslctx_setup(void* ctxt, const char* tls_protocols)
+listen_sslctx_setup(void* ctxt, const char* tls_protocols,
+	int use_system_versions)
 {
 #ifdef HAVE_SSL
 	int allow12, allow13;
 	SSL_CTX* ctx = (SSL_CTX*)ctxt;
 	cfg_tls_protocols_allowed(tls_protocols, &allow12, &allow13);
 
+	if(!use_system_versions) {
 	/* no SSLv2, SSLv3 because has defects */
 #if SSL_OP_NO_SSLv2 != 0
 	if((SSL_CTX_set_options(ctx, SSL_OP_NO_SSLv2) & SSL_OP_NO_SSLv2)
@@ -1282,6 +1284,7 @@ listen_sslctx_setup(void* ctxt, const char* tls_protocols)
 		}
 	}
 #endif
+	}
 #if defined(SSL_OP_NO_RENEGOTIATION)
 	/* disable client renegotiation */
 	if((SSL_CTX_set_options(ctx, SSL_OP_NO_RENEGOTIATION) &
@@ -1320,7 +1323,7 @@ listen_sslctx_setup(void* ctxt, const char* tls_protocols)
 	SSL_CTX_set_security_level(ctx, 0);
 #endif
 #else
-	(void)ctxt; (void)tls_protocols;
+	(void)ctxt; (void)tls_protocols; (void)use_system_versions;
 #endif /* HAVE_SSL */
 	return 1;
 }
@@ -1356,7 +1359,8 @@ listen_sslctx_setup_2(void* ctxt)
 void* listen_sslctx_create(const char* key, const char* pem,
 	const char* verifypem, const char* tls_ciphers,
 	const char* tls_ciphersuites, int set_ticket_keys_cb,
-	int is_dot, int is_doh, const char* tls_protocols)
+	int is_dot, int is_doh, const char* tls_protocols,
+	int use_system_versions)
 {
 #ifdef HAVE_SSL
 	SSL_CTX* ctx = SSL_CTX_new(SSLv23_server_method());
@@ -1374,7 +1378,7 @@ void* listen_sslctx_create(const char* key, const char* pem,
 		SSL_CTX_free(ctx);
 		return NULL;
 	}
-	if(!listen_sslctx_setup(ctx, tls_protocols)) {
+	if(!listen_sslctx_setup(ctx, tls_protocols, use_system_versions)) {
 		SSL_CTX_free(ctx);
 		return NULL;
 	}
diff --git a/util/net_help.h b/util/net_help.h
index e9f6f35..58a2e76 100644
--- a/util/net_help.h
+++ b/util/net_help.h
@@ -479,9 +479,12 @@ void log_cert(unsigned level, const char* str, void* cert);
  * Set SSL_OP_NOxxx options on SSL context to disable bad crypto
  * @param ctxt: SSL_CTX*
  * @param tls_protocols: configure string with allowed TLS protocols to use.
+ * @param use_system_versions: rely on the system policy (if any) for allowed
+ *	TLS versions
  * @return false on failure.
  */
-int listen_sslctx_setup(void* ctxt, const char* tls_protocols);
+int listen_sslctx_setup(void* ctxt, const char* tls_protocols,
+	int use_system_versions);
 
 /**
  * Further setup of listening SSL context, after keys loaded.
@@ -501,12 +504,15 @@ void listen_sslctx_setup_2(void* ctxt);
  * @param is_dot: if the TLS connection is for DoT to set the appropriate ALPN.
  * @param is_doh: if the TLS connection is for DoH to set the appropriate ALPN.
  * @param tls_protocols: configure string with allowed TLS protocols to use.
+ * @param use_system_versions: rely on the system policy (if any) for allowed
+ *	TLS versions
  * return SSL_CTX* or NULL on failure (logged).
  */
 void* listen_sslctx_create(const char* key, const char* pem,
 	const char* verifypem, const char* tls_ciphers,
 	const char* tls_ciphersuites, int set_ticket_keys_cb,
-	int is_dot, int is_doh, const char* tls_protocols);
+	int is_dot, int is_doh, const char* tls_protocols,
+	int use_system_versions);
 
 /**
  * create SSL connect context
diff --git a/winrc/win_svc.c b/winrc/win_svc.c
index c0792ce..efc50d3 100644
--- a/winrc/win_svc.c
+++ b/winrc/win_svc.c
@@ -369,7 +369,8 @@ service_init(int r, struct daemon** d, struct config_file** c)
 			cfg->tls_ciphers, cfg->tls_ciphersuites,
 			(cfg->tls_session_ticket_keys.first &&
 			cfg->tls_session_ticket_keys.first->str[0] != 0),
-			1, 0, cfg->tls_protocols))) {
+			1, 0, cfg->tls_protocols,
+			cfg->tls_use_system_policy_versions))) {
 			fatal_exit("could not set up listen SSL_CTX");
 		}
 #ifdef HAVE_NGHTTP2_NGHTTP2_H
@@ -379,7 +380,8 @@ service_init(int r, struct daemon** d, struct config_file** c)
 				cfg->tls_ciphers, cfg->tls_ciphersuites,
 				(cfg->tls_session_ticket_keys.first &&
 				cfg->tls_session_ticket_keys.first->str[0] != 0),
-				0, 1, cfg->tls_protocols))) {
+				0, 1, cfg->tls_protocols,
+				cfg->tls_use_system_policy_versions))) {
 				fatal_exit("could not set up listen doh SSL_CTX");
 			}
 		}
-- 
2.54.0

